src/runtime/runtime.ts

```diff
@@ -221,12 +253,81 @@ export class Runtime {
     // we could eliminate it if the Manifest object takes care of this.
     const id = `in-memory-${Math.floor((Math.random()+1)*1e6)}.manifest`;
     // TODO(sjmiles): this is a virtual manifest, the fileName is invented
-    const localOptions = {id, fileName: `./${id}`, loader};
-    return Manifest.parse(content, {...localOptions, ...options});
+    const opts = {id, fileName: `./${id}`, loader, ...options};
+    return Manifest.parse(content, opts);
+  }
+
+  async parseFile(path: string, options?): Promise<Manifest> {
+    const content = await this.loader.loadResource(path);
+    const opts = {id: path, fileName: path, loader: this.loader, ...options};
+    return this.parse(content, opts);
+  }
+
+  async resolveRecipe(arc: Arc, recipe: Recipe): Promise<Recipe | null> {
+    if (this.normalize(recipe)) {
+      if (recipe.isResolved()) {
+        return recipe;
+      }
+      const resolver = new RecipeResolver(arc);
+      const plan = await resolver.resolve(recipe);
+      if (plan && plan.isResolved()) {
+        return plan;
+      }
+      warn('failed to resolve:\n', (plan || recipe).toString({showUnresolved: true}));
+    }
+    return null;
   }
 
+  normalize(recipe: Recipe): boolean {
+    if (Runtime.isNormalized(recipe)) {
+      return true;
+    }
+    const errors = new Map();
+    if (recipe.normalize({errors})) {
+      return true;
+    }
+    warn('failed to normalize:\n', errors, recipe.toString());
+    return false;
+  }
+
+  static isNormalized(recipe: Recipe): boolean {
+    return Object.isFrozen(recipe);
+  }
+
+  // TODO(sjmiles): redundant vs. newArc, but has some impedance mismatch
```

Comment:
There's actually three seemingly equivalent methods: runArc, newArc and spawnArc. Consolidating them would be great. Until that happens though, documenting the difference between the methods would be really helpful.

Reply:
Will do.

Reply:
The differences are subtle, I'm not sure I can coherently define them in code comments (apart from just restating the different argument structures). 

I made a follow up ticket #4070